test/hotspot/gtest/runtime/test_os.cpp

```diff
@@ -551,26 +553,29 @@ TEST_VM(os, release_bad_ranges) {
 TEST_VM(os, release_one_mapping_multi_commits) {
   // Test that we can release an area consisting of interleaved
   //  committed and uncommitted regions:
-  const size_t stripe_len = 4 * M;
-  const int num_stripes = 4;
+  const size_t stripe_len = os::vm_allocation_granularity();
+  const int num_stripes = 6;
   const size_t total_range_len = stripe_len * num_stripes;
 
   // reserve address space...
   address p = reserve_one_commit_multiple(num_stripes, stripe_len);
-  ASSERT_NE(p, (address)NULL);
   PRINT_MAPPINGS("A");
+  ASSERT_NE(p, (address)nullptr);
 
-  // .. release it...
-  ASSERT_TRUE(os::release_memory((char*)p, total_range_len));
+  // reserve at border of the region
+  address border = p + num_stripes * stripe_len;
+  address p2 = (address)os::attempt_reserve_memory_at((char*)border, stripe_len, false);
   PRINT_MAPPINGS("B");
 
-  // re-reserve it. This should work unless release failed.
-  address p2 = (address)os::attempt_reserve_memory_at((char*)p, total_range_len);
-  ASSERT_EQ(p2, p);
-  PRINT_MAPPINGS("C");
+  ASSERT_TRUE(p2 == nullptr || p2 == border);
```

Comment:
Why are we allowing this case to fail?

Reply:
There is no guarantee that the attempt will always be successful. By 'attempt', we mean that failing to reserve is also expected.

Reply:
Yes, something does not match up.

The original test did this:

1) reserve a stripy region
2) release the stripy region
3) re-reserve in the same area of the stripy region. If the re-reserve worked, that was a proof that the release worked too, because we have now an address hole.

The problem with the original test was that (3) could fail because some other concurrent mmap got placed into that just released address space hole. That would have been okay, but we cannot tell apart: A) the release in step (2) did fail and we actually still have the original reservation in that address hole, which would be an error. Or, B) the release in step (2) did succeed, and some concurrent thread grabbed the hole now, which would probably be okay.

There are two ways to fix this:

- we either remove the (3) re-reservation. This means essentially giving up, which is okay in my eyes considering how much time this did cost already.
- my proposal was to place border regions at the ends of the stripy region, before releasing it:

1) reserve a stripy region
1.5) place border region before and after region
2) release the stripy region. Now the address hole is limited at both sides, which drastically reduces the chance of concurrent allocations getting into that hole. Especially since you did reduce the stripe size, which is good.
3) re-reserve in the same area of the stripy region. That should work, hopefully, so often that we can consider the test to be safe.

You did both - placed a border region (only one), but also removed the re-reservation. If you remove the re-reservation, you don't have to do the border region.

Reply:
Looking at this I am starting to wonder whether we really need this test at all.

We seem to be testing whether the underlaying OS does what we want it, as opposed to our VM code, which in this case is just a thin convenience layer.

I can see this test having been useful when we were first implementing/using these features, but what would we need to get wrong nowadays to cause a regression in this test for example, short of someone doing a complete rewrite and getting something wrong?

Reply:
> Looking at this I am starting to wonder whether we really need this test at all.
> 
> We seem to be testing whether the underlaying OS does what we want it, as opposed to our VM code, which in this case is just a thin convenience layer.
> 
> I can see this test having been useful when we were first implementing/using these features, but what would we need to get wrong nowadays to cause a regression in this test for example, short of someone doing a complete rewrite and getting something wrong?

I wrote this originally (I believe) because of a very real issue where, on Windows, we would allocate memory in stripes for different NUMA nodes, and then later release the memory. IIRC, this just released the first stripe, so the rest stayed allocated (os::release_memory(p, size) called VirtualFree(MEM_RELEASE, p), ignoring the size parameter). 

We then fixed os::release_memory to, if called over multiple reservations, release them all - query all existing reservations in that area with VirtualQuery, then releasing them. That stuff is intricate and it can fail, so a regression test makes actually a lot of sense.

Reply:
Can we guard (using `#if`, for example) this test to run only in Windows platform?

Reply:
> Can we guard (using `#if`, for example) this test to run only in Windows platform?

Since we must do it anyhow on Windows with NUMA, I feel like we might as well deploy it on all platforms, just in case, but it is a valid suggestion.